[GTK][WPE] Use one fbo per render target instead of a global one in AcceleratedSurfaceDMABuf

carlosgcampos opened the pull request:
#### 3c971961dbb013fd56c20ce20fee776f4c89ccee
<pre>
[GTK][WPE] Use one fbo per render target instead of a global one in AcceleratedSurfaceDMABuf
<a href="https://bugs.webkit.org/show_bug.cgi?id=287951">https://bugs.webkit.org/show_bug.cgi?id=287951</a>

Reviewed by Nikolas Zimmermann.

Instead of changing the attachments of the fbo every time we render a
frame, we have one fbo per target where attachments are set once on
construction and then we just switch the current fbo before rendering a
new frame. This patch removes the intermediate class
RenderTargetColorBuffer and both RenderTargetEGLImage and
RenderTargetSHMImage create its own color buffer and attach it to the
fbo.

* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::RenderTarget):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::~RenderTarget):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::setReleaseFenceFD):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::~RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::RenderTargetSHMImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::~RenderTargetSHMImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::RenderTargetTexture):
(WebKit::AcceleratedSurfaceDMABuf::willDestroyGLContext):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::waitRelease): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::RenderTargetColorBuffer): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::~RenderTargetColorBuffer): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::didCreateGLContext): Deleted.
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:

Co-authored-by: Nikolas Zimmermann &lt;nzimmermann@igalia.com&gt;
Canonical link: <a href="https://commits.webkit.org/290604@main">https://commits.webkit.org/290604@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1170c84c9e7c2a9d95ae0ce0856dc6351f801635

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90609 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95621 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41391 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10538 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18460 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93610 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/50077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40521 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/37617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97450 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17801 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18059 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22389 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/21013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14256 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17811 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17550 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21005 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19334 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (approved):
Looks good.

aperezdc commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -115,16 +115,25 @@ AcceleratedSurfaceDMABuf::RenderTarget::RenderTarget(uint64_t surfaceID, const W
 
+    if (m_fbo)
+        glDeleteFramebuffers(1, &m_fbo);
+
```
We should destroy the FBO before the depth/stencil buffer attached to it.

```suggestion
    if (m_fbo)
        glDeleteFramebuffers(1, &m_fbo);

    if (m_depthStencilBuffer)
        glDeleteRenderbuffers(1, &m_depthStencilBuffer);

```

aperezdc commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
```diff
@@ -125,24 +124,19 @@ class AcceleratedSurfaceDMABuf final : public AcceleratedSurface, public IPC::Me
 #endif
 
-        virtual void willRenderFrame() const;
-        virtual void didRenderFrame()
```
Very nice being able to remove a virtual method 💪🏼

carlosgcampos committed:
[GTK][WPE] Use one fbo per render target instead of a global one in AcceleratedSurfaceDMABuf
https://bugs.webkit.org/show_bug.cgi?id=287951

Reviewed by Nikolas Zimmermann.

Instead of changing the attachments of the fbo every time we render a
frame, we have one fbo per target where attachments are set once on
construction and then we just switch the current fbo before rendering a
new frame. This patch removes the intermediate class
RenderTargetColorBuffer and both RenderTargetEGLImage and
RenderTargetSHMImage create its own color buffer and attach it to the
fbo.

* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::RenderTarget):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::~RenderTarget):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::setReleaseFenceFD):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetEGLImage::~RenderTargetEGLImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::RenderTargetSHMImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::~RenderTargetSHMImage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::RenderTargetTexture):
(WebKit::AcceleratedSurfaceDMABuf::willDestroyGLContext):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::waitRelease): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::RenderTargetColorBuffer): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::~RenderTargetColorBuffer): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetTexture::willRenderFrame const): Deleted.
(WebKit::AcceleratedSurfaceDMABuf::didCreateGLContext): Deleted.
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:

Co-authored-by: Nikolas Zimmermann <nzimmermann@igalia.com>
Canonical link: https://commits.webkit.org/290604@main
